### PR TITLE
[Merged by Bors] - Add methods for querying lists of entities.

### LINKED ISSUE #1470: Ordered iteration over Queries
**What problem does this solve or what need does it fill?**

Iterating over queries in a particular order is a reasonably common pattern: it is used for a) prioritization and early stopping b) ensuring reproducible execution, particularly when dealing with PRNG or modifying a common piece of data with floating point math.

It can be done manually with the current tools, but they are frustrating to write and read, and hard to optimize.

**What solution would you like?**

Provide a pair of `Query` methods: `ordered_iter` and `ordered_iter_mut`, which iterate over the entities in the query in the specified order.

In the example demonstrating this feature, show how to iterate in `Entity` order, and explain why using `Entity` as your ordering is / isn't stable, because it's the obvious thing to try and that property relies on deep engine internals.

In the example demonstrating this feature, show how to sort your entities by the value of one of their components, then iterate over the query in that order.

**What alternative(s) have you considered?**

Manually implement this functionality in each system using a for loop over the ordering, then using the appropriate `query.get` call.

Manually implement this functionality as a method on `Query` using a trait in each app that needs this functionality.

I'm not a fan of either of these approaches, as a) they create a lot of conceptual overhead and b) getting an optimized iteration method is challenging from the outside.

**Additional context**

The `ordered_iter` crate has a nice reference implementation / [API](https://docs.rs/ordered_iter/0.1.2/ordered_iter/trait.OrderedSetIterator.html) that we can probably borrow from.

This may be efficient to implement at the same time as #762.

It's not immediately clear how / if to combine this with parallel iteration. It's useful in some cases, but nonsensical in others.

### PR DESCRIPTION
**Pull request by devil-ira**
# Objective
Improve querying ergonomics around collections and iterators of entities.

Example how queries over Children might be done currently. 
```rust
fn system(foo_query: Query<(&Foo, &Children)>, bar_query: Query<(&Bar, &Children)>) {
    for (foo, children) in &foo_query {
        for child in children.iter() {
            if let Ok((bar, children)) = bar_query.get(*child) {
                for child in children.iter() {
                    if let Ok((foo, children)) = foo_query.get(*child) {
                        // D:
                    }
                }
            }
        }
    }
}
```
Answers #4868
Partially addresses #4864
Fixes #1470
## Solution
Based on the great work by @deontologician in #2563 

Added `iter_many` and `many_for_each_mut` to `Query`.
These take a list of entities (Anything that implements `IntoIterator<Item: Borrow<Entity>>`).

`iter_many` returns a `QueryManyIter` iterator over immutable results of a query (mutable data will be cast to an immutable form).

`many_for_each_mut` calls a closure for every result of the query, ensuring not aliased mutability. 
This iterator goes over the list of entities in order and returns the result from the query for it. Skipping over any entities that don't match the query.

Also added `unsafe fn iter_many_unsafe`.

### Examples
```rust
#[derive(Component)]
struct Counter {
    value: i32
}

#[derive(Component)]
struct Friends {
    list: Vec<Entity>,
}

fn system(
    friends_query: Query<&Friends>,
    mut counter_query: Query<&mut Counter>,
) {
    for friends in &friends_query {
        for counter in counter_query.iter_many(&friends.list) {
            println!("Friend's counter: {:?}", counter.value);
        }
        
        counter_query.many_for_each_mut(&friends.list, |mut counter| {
            counter.value += 1;
            println!("Friend's counter: {:?}", counter.value);
        });
    }
}

```

Here's how example in the Objective section can be written with this PR.
```rust
fn system(foo_query: Query<(&Foo, &Children)>, bar_query: Query<(&Bar, &Children)>) {
    for (foo, children) in &foo_query {
        for (bar, children) in bar_query.iter_many(children) {
            for (foo, children) in foo_query.iter_many(children) {
                // :D
            }
        }
    }
}
```
## Additional changes
Implemented `IntoIterator` for `&Children` because why not.
## Todo
- Bikeshed!

Co-authored-by: deontologician <deontologician@gmail.com>

### CODE REVIEW

**Comment by cart** [comment]
Alternatively, could we remove the need for an extension trait by putting this on the Query impl?

```rust
impl<BLAH> Query<BLAH> {
  fn iter_with<ENTITY_ITER>(&self, i: ENTITY_ITER) -> QueryIter<BLAH> {
  }
}


fn system(has_a: Query<Entity, With<A>>, bs: Query<&B>) {
   for b in bs.iter_with(&has_a) {
   }
}
```

Functionally, this is very similar to the `many` api variants. The only difference is whether or not we use const-length-arrays vs iterators. We should definitely support both as they have different uses, but we might want to unify terminology. Maybe `many_iter`, `many_for_each`, and `many_for_each_mut`?

**Comment by james7132** [comment]
> Functionally, this is very similar to the `many` api variants. The only difference is whether or not we use const-length-arrays vs iterators. We should definitely support both as they have different uses, but we might want to unify terminology. Maybe `many_iter`, `many_for_each`, and `many_for_each_mut`?

The main use case for `many`'s variants is *simultaneous* mutative access to multiple fetched results, which incurs a `O(k^2)` checks to ensure aliased access is not done. This may not be possible without iterators that can be seeked/restarted.

**Comment by cart** [comment]
> The main use case for many's variants is simultaneous mutative access to multiple fetched results, which incurs a O(k^2) checks to ensure aliased access is not done. This may not be possible without iterators that can be seeked/restarted.

Yup I'm aware. The collision check is only for many_mut. The iterator in this PR also allows simultaneous access and also can skip the collision check (but only because it limits itself to read-only access, just like many_mut). They are literally the same behavior and effects, just different apis / language features.

File: crates/bevy_ecs/src/query/iter.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/iter.rs:471]: I'd prefer to use standard database terms. This is an "inner join". We could add left, right and outer joins too, which would be nice.

They'd require making some fields options though.

File: crates/bevy_ecs/src/query/iter.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/iter.rs:481]: I don't love these methods, especially with such confusing names. Why are they better than just doing a join and then a for_each?

File: crates/bevy_hierarchy/src/components/children.rs
Comment (alice-i-cecile) [inline crates/bevy_hierarchy/src/components/children.rs:46]: This change is super useful and uncontroversial. I love it.

**Comment by cart** [comment]
Minor correction to what I wrote above:

> (but only because it limits itself to read-only access, just like many_mut). 

`many_mut` clearly doesn't limit itself to read-only access. I meant to say `many`. `many` vs `many_iter` (or whatever we decide to call it) _are_ the same behavior, just different apis / const-ness.

**Comment by alice-i-cecile** [comment]
@devil-ira can we use this to solve #1430 too?

**Comment by devil-ira** [comment]
Looks like it. And @james7132 had some ideas about optimizing these methods. Maybe in a follow-up pr?

File: crates/bevy_ecs/src/query/iter.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/iter.rs:415]: Potential optimization: unpack `Query::get` and cache the required parts inside the iterator.

 - Use `QueryState` over `Query`
 - initialize the fetches on construction (like QueryIter)
 - cache a reference to `Entities`
 - Duplicate the same checks as `Query::get` but with the cached references instead.

If this seems like a lot, I can probably do this in a followup PR.

**Review by james7132 — APPROVED** [review]
Generally looks good for now. There is an opportunity for optimization here that is not in `Query::many`, but as mentioned above it can be deferred to a later PR. 

I actually prefer this over `Query::many` and it's variants due to the composability with other iterators, which is distinctly more Rust like, and I'm not sure what contexts would prefer a compile-time array version over iterators in typical Bevy code. This is also generally inside the engine as a first party user of this API: transform propagation could definitely use this and benefit from the optimization.

A mut version of this could be done with a `HashSet<Entity>` or something similar, panicking when an entity that was previously seen is encountered again, but that has the potential to allocate memory (might be worth using something like `tinyset`?).

**Comment by alice-i-cecile** [comment]
> and I'm not sure what contexts would prefer a compile-time array version over iterators in typical Bevy code

This is actually quite useful when considering interactions between e.g. pairs of entities. But yes, I expect this API to be much faster and more composable for n > 2.

**Comment by devil-ira** [comment]
I've added `many_for_each_unchecked_manual` to `QueryState`.
Internally it's similar to `get_unchecked_manual` so the new unsafe code isn't *too* scary.

I'll work on `QueryManyIter` next. Which should improve performance and make `many_iter_mut` possible.

**Comment by devil-ira** [comment]
*(Rebased because of #4665 so my local CI matches the live CI)*

**Comment by devil-ira** [comment]
Yeah, previously I wasn't sure how to make `many_iter_mut` work. So I kept the `for_each` variants around.
Just realized myself they can be yoten.

File: crates/bevy_ecs/src/query/iter.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/iter.rs:163]: This will fail when it cannot find an entity, we need it to skip over them.

File: crates/bevy_ecs/src/query/iter.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/iter.rs:156]: Can you use a `&'w` reference to  all three of these instead of in `next`? No need to derefence a `&'w World` in a hot loop when you can avoid it.

File: crates/bevy_ecs/src/query/iter.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/iter.rs:194]: Shouldn't this defer to `entity_iter.size_hint()`?

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/system/query.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/query.rs:1206]:
```suggestion
    ///  # Safety
    ///  This does not check for entity uniqueness.
    ///  The caller must never produce aliased mutability of component data belonging to the same entity.
    ///  Particular care must be taken when collecting the data (rather than iterating over it one item at a time)
    /// such as via `[Iterator::collect()]`.
    pub unsafe fn many_iter_mut<EntityList: IntoIterator>(
```

**Comment by alice-i-cecile** [comment]
> we could make a safe variant that accepts only Children and Query<Entity>, but maybe that's too particular

This makes safety assumptions in very far-flung places in ways that make me nervous, so I'd be against it :)

**Comment by devil-ira** [comment]
I'm really starting to like the idea of making `many_iter_mut` safe, but restricting the entities parameter to ensure no overlapping mutable fetches happen.
How about an unsafe trait that we call `EntityListDeduped` or something.
Then we implement this trait on Children, Query\<Entity\> and HashSet\<Entity\>. Essentially, on anything that ensures each entity is unique.
With this, and some well written docs it should be very usable, I believe.

**Comment by devil-ira** [comment]
@alice-i-cecile I totally agree that we shouldn't *just* rely on Children being unique.
It would need an unsafe trait impl, strong tests and a solid safety comment.

But I suppose that's only going to happen if James was planning on ensuring correctness of the heirarchy in this way. I'm not exactly sure what the plan is.

But Children aside. Would you be okay with HashSet and Query? They both make absolute guarantees about uniqueness of the entities.

**Comment by alice-i-cecile** [comment]
>  Would you be okay with HashSet and Query? They both make absolute guarantees about uniqueness of the entities.

I would be 100% okay with this. I would still like the unsafe version to exist, but feel free to bikeshed about which one gets the prestigious iter_mut() name.

**Comment by cart** [comment]
I almost want to just replace it with for_each_mut and call it a day. Internally allocating + hashing (when thats the only method in query that does it) seems like a foot-gun, especially when the immutable variant doesn't do it.

File: crates/bevy_ecs/src/query/iter.rs
Comment (james7132) [inline crates/bevy_ecs/src/query/iter.rs:149]: nit: probably more compact as a match statement.

**Comment by devil-ira** [comment]
I've gone with what cart and alice have suggested.
- `many_for_each_mut` has been added back.
- Decided to rename `many_iter` to `iter_many` as it looked out of place next to `iter*` and `iter_combinations*` and co.
- `many_iter_mut` is now `unsafe fn iter_many_unsafe` following the style of `iter_unsafe`.

File: crates/bevy_ecs/src/system/query.rs
Comment (cart) [inline crates/bevy_ecs/src/system/query.rs:669]: I just ensured that data isn't allowed to escape the closure (which is definitely what we want). To ensure we never regress here (ex: by changing this to `QueryItem<'w, Q>`), we should add a new "compile fail" test to `bevy_ecs_compile_fail_tests`. Check out tests like `system_query_get_many_lifetime_safety` for an example.

This is the type of thing that should never be allowed to compile, as it essentially turns this into a mutable iterator:

```rust
let mut results = Vec::new();
query.many_for_each_mut(vec![e, e], |a| {
    results.push(a);
});
```

**Review by cart — COMMENTED** [review]
I think this is pretty much good to go, once we add the relevant "safety regression" test (see the comment below).

Also, ideally we also have safe variants on QueryState (see the equivalents for things like `query_state.iter()`).

**Review by alice-i-cecile — APPROVED** [review]

**Review by james7132 — APPROVED** [review]

**Pull request closed without merging**
